File: src/utils/datasets_io.R
read_datasets_json <- function(path = "datasets.json", view = NULL) {
  datasets <- jsonlite::fromJSON(path, simplifyVector = FALSE)

  result <- list()

  for (ds_name in names(datasets)) {
    ds <- datasets[[ds_name]]
    views <- ds[["views"]]
    base_columns <- ds[["columns"]]
    if (is.null(base_columns)) base_columns <- list()

    if (is.null(views)) next

    matched_views <- list()
    for (v_name in names(views)) {
      if (!is.null(view) && v_name != view) next

      v <- views[[v_name]]
      output_file <- v[["output_file_name"]]
      if (is.null(output_file)) next

      view_columns <- v[["columns"]]
      if (is.null(view_columns)) view_columns <- list()
      matched_views[[v_name]] <- list(
        view_name = v_name,
        input_file = v[["input_file_name"]],
        output_file = output_file,
        subset_vars = v[["subset_vars"]],
        columns = modifyList(base_columns, view_columns)
      )
    }

    if (length(matched_views) == 0) next

    first_v_name <- names(matched_views)[1]
    first_v <- matched_views[[first_v_name]]
    columns <- first_v[["columns"]]

    entry <- list(
      # dataset-level fields (order mirrors datasets.json)
      display_name = ds[["display_name"]],
      file_names = ds[["file_names"]],
      folder_name = ds[["folder_name"]],
      tissue = ds[["tissue"]],
      normal_tissue = ds[["normal_tissue"]],
      use_for_benchmark = ds[["use_for_benchmark"]],
      use_for_batch_effect = ds[["use_for_batch_effect"]],
      # columns (order mirrors datasets.json "columns")
      sample_col = columns[["sample"]],
      label_col = columns[["label"]],
      batch_col = columns[["batch"]],
      cell_type_low_res = columns[["cell_type_low_res"]],
      cell_type_high_res = columns[["cell_type_high_res"]],
      # Optional flat flag: annotation methods this dataset is NOT suitable
      # for (values: "hitme", "scatomic"). Absent/empty = suitable for all
      # (see AGENTS.md "Onboarding new datasets" / benchmark_pipeline.R).
      not_suitable_for_auto_annotation = if (is.null(ds[["not_suitable_for_auto_annotation"]])) {
        character(0)
      } else {
        ds[["not_suitable_for_auto_annotation"]]
      },
      # first-matching-view summary (R-compat; mirrors view entry order)
      view_name = first_v_name,
      input_file = first_v[["input_file"]],
      output_file = first_v[["output_file"]],
      subset_vars = first_v[["subset_vars"]],
      views = matched_views
    )

    result[[ds_name]] <- entry
  }

  return(result)
}

# Resolve the preprocessed view h5ad file name for a dataset (mirrors
# get_h5ad_path in benchmark_hpc_utils.R, but returns the bare file name —
# the notebook composes the full path from its own NAS mount). `datasets` is
# the read_datasets_json(view = ...) output; `ds` the datasets.json key.
get_view_h5ad_path <- function(datasets, ds, view) {
  entry <- datasets[[ds]]
  if (is.null(entry)) {
    stop("Dataset '", ds, "' not found in datasets.json (view '", view, "').")
  }
  views <- entry[["views"]]
  if (is.null(views) || is.null(views[[view]])) {
    stop("Dataset '", ds, "' has no '", view, "' view in datasets.json.")
  }
  out_file <- views[[view]][["output_file"]]
  if (is.null(out_file) || is.na(out_file) || out_file == "") {
    stop("Dataset '", ds, "' view '", view,
         "' has no output_file_name in datasets.json.")
  }
  return(out_file)
}

File: src/utils/py/datasets_io.py
"""Python analog of src/utils/datasets_io.R::read_datasets_json().

Stdlib-only (json) so it can be imported from anywhere, including HPC
worker scripts that must not pull in R (rpy2) or heavy analysis deps.
"""

import json


def read_datasets_json(path="datasets.json", view=None):
    """Read datasets.json into a dict keyed by dataset name.

    Returns {ds_name: entry} with ALL datasets.json fields (future-proof).
    Entry summary fields (view_name, output_file, input_file, subset_vars)
    refer to the FIRST view matching `view` (with an output_file_name),
    mirroring the R version. `views` holds ALL matching views. Datasets
    without any matching view (e.g. Zhu, view-less raw source) are included
    with summary fields set to None/{} so raw inputs (file_names) are
    discoverable; the R version skips them (its callers use view filters).
    """
    with open(path) as f:
        datasets = json.load(f)

    result = {}
    for ds_name, ds in datasets.items():
        views = ds.get("views") or {}

        matched_views = {}
        for v_name, v in views.items():
            if view is not None and v_name != view:
                continue
            output_file = v.get("output_file_name")
            if not output_file:
                continue
            matched_views[v_name] = {
                "input_file": v.get("input_file_name"),
                "output_file": output_file,
                "subset_vars": v.get("subset_vars", {}),
            }

        first_name = next(iter(matched_views), None)
        first = matched_views.get(first_name)
        columns = ds.get("columns") or {}

        result[ds_name] = {
            # dataset-level fields (order mirrors datasets.json)
            "display_name": ds.get("display_name"),
            "file_names": ds.get("file_names"),
            "folder_name": ds.get("folder_name"),
            "tissue": ds.get("tissue"),
            "normal_tissue": ds.get("normal_tissue"),
            "use_for_benchmark": ds.get("use_for_benchmark"),
            "use_for_batch_effect": ds.get("use_for_batch_effect"),
            # columns (order mirrors datasets.json "columns")
            "sample_col": columns.get("sample"),
            "label_col": columns.get("label"),
            "batch_col": columns.get("batch"),
            "cell_type_low_res": columns.get("cell_type_low_res"),
            "cell_type_high_res": columns.get("cell_type_high_res"),
            "meta_cols_keep": ds.get("meta_cols_keep"),
            # first-matching-view summary (R-compat; mirrors view entry order)
            "view_name": first_name,
            "input_file": first["input_file"] if first else None,
            "output_file": first["output_file"] if first else None,
            "subset_vars": first["subset_vars"] if first else {},
            # all matching views
            "views": matched_views,
        }

    return result
